File: BAEKJOON/java/13904_과제.java
import java.io.*;
import java.util.*;

public class Main {
    public static int n, max_day = -1, answer = 0;
    public static Map<Integer, ArrayList<Integer>> map = new HashMap<>();
    public static List<Assignment> list = new ArrayList<>();
    public static List<Integer> task = new ArrayList<>();

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        input();
        solution();

        bw.flush();
        bw.close();
    }

    private static void input() throws IOException {
        n = Integer.parseInt(br.readLine());

        StringTokenizer st;

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (max_day < a) {
                max_day = a;
            }

            if (map.containsKey(a)) {
                map.get(a).add(b);
                continue;
            }

            if (!map.containsKey(a)) {
                map.put(a, new ArrayList<>(Arrays.asList(b)));
            }
        }
    }

    private static void solution() throws IOException {
        while (max_day > 0) {
            if (map.containsKey(max_day)) {
                for (int i = 0; i < map.get(max_day).size(); i++) {
                    Assignment assignment = new Assignment(map.get(max_day).get(i), max_day);

                    list.add(assignment);
                }
            }

            Collections.sort(list);

            if (list.isEmpty()) {
                max_day--;
                continue;
            }

            task.add(list.get(0).score);

            list.remove(0);

            max_day--;
        }

        task.stream().forEach(x -> answer += x);

        System.out.println(answer);
    }
}

class Assignment implements Comparable<Assignment> {
    int score;
    int RemainingDays;

    public Assignment(int score, int remainingDays) {
        this.score = score;
        RemainingDays = remainingDays;
    }

    @Override
    public int compareTo(Assignment o) {
        if (this.score < o.score) {
            return 1;
        }

        if (this.score == o.score) {
            return 0;
        }

        return -1;
    }
}

File: BAEKJOON/cpp/13904_과제.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

int n, max_day = -1, answer = 0;
unordered_map<int, vector<int>> um; //남은 일수, 점수
vector<pair<int, int>> v; //점수, 남은 일수
vector<int> task;

void input() {
	cin >> n;

	int a, b;
	for (int i = 0; i < n; i++) {
		cin >> a >> b;
		um[a].push_back(b);

		if (max_day < a) {
			max_day = a;
		}
	}
}

void solution() {
	while (max_day > 0) {
		for (int i = 0; i < um[max_day].size(); i++) {
			v.push_back({ um[max_day][i], max_day });
		}

		sort(v.begin(), v.end(), greater<>());
		
		if (v.empty()) {
			max_day--;
			continue;
		}

		task.push_back(v[0].first);

		v.erase(v.begin(), v.begin() + 1);

		max_day--;
	}

	for (int i = 0; i < task.size(); i++) {
		answer += task[i];
	}

	cout << answer;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	input();
	solution();

	return 0;
}